### Experimenting by removing Upvote and Downvote columns
Now we will experiment by removing Upvote and Downvote and then see the MAPE results.

In [1]:
# Importing Necessary Modules
import pandas as pd
import numpy as np
import spacy
nlp=spacy.load('en_core_web_sm')
from sklearn.model_selection import train_test_split
from spacy.pipeline import SentenceSegmenter
import re
import warnings
warnings.filterwarnings('ignore')
import textblob
from textblob import TextBlob
from spacy.lang.en import English

# Customization of View
pd.options.display.max_columns = 999

In [2]:
# Importing the dataset containing unigram feature set
df = pd.read_csv('Unigram_Feature.csv',index_col='Sr_No')

### Getting an understanding of the data

In [3]:
df.head()

Unnamed: 0  1080p  12mp  1960mah  32gb  3gs  44000rs  500rs  59k  720p  \
Sr_No                                                                           
1               0    0.0   0.0      0.0   0.0  0.0      0.0    0.0  0.0   0.0   
2               1    0.0   0.0      0.0   0.0  0.0      0.0    0.0  0.0   0.0   
3               2    0.0   0.0      0.0   0.0  0.0      0.0    0.0  0.0   0.0   
5               3    0.0   0.0      0.0   0.0  0.0      0.0    0.0  0.0   0.0   
7               4    0.0   0.0      0.0   0.0  0.0      0.0    0.0  0.0   0.0   

       7mp  8pm    a  a10  aa10  able  absence  absolutely  accelerated  \
Sr_No                                                                     
1      0.0  0.0  0.0  0.0   0.0   0.0      0.0         0.0          0.0   
2      0.0  0.0  0.0  0.0   0.0   0.0      0.0         0.0          0.0   
3      0.0  0.0  0.0  0.0   0.0   0.0      0.0         0.0          0.0   
5      0.0  0.0  0.0  0.0   0.0   0.0      0.0         0.0          0.0   
7      0.0  0.0  0.0  0.0   0.0   0.0      0.0         0.0          0.0   

       accept  accesorie  access  accessory  accord  accumulation  achieved  \
Sr_No                                                                         
1         0.0        0.0     0.0        0.0     0.0           0.0       0.0   
2         0.0        0.0     0.0        0.0     0.0           0.0       0.0   
3         0.0        0.0     0.0        0.0     0.0           0.0       0.0   
5         0.0        0.0     0.0        0.0     0.0           0.0       0.0   
7         0.0        0.0     0.0        0.0     0.0           0.0       0.0   

       acknowledge  act  activate  activity    actual   ad  adapt  add  \
Sr_No                                                                    
1              0.0  0.0       0.0       0.0  0.000000  0.0    0.0  0.0   
2              0.0  0.0       0.0       0.0  0.194899  0.0    0.0  0.0   
3              0.0  0.0       0.0       0.0  0.000000  0.0    0.0  0.0   
5              0.0  0.0       0.0       0.0  0.000000  0.0    0.0  0.0   
7              0.0  0.0       0.0       0.0  0.000000  0.0    0.0  0.0   

       additional  address  adjust  admit  advance  advantage  advocate  \
Sr_No                                                                     
1             0.0      0.0     0.0    0.0      0.0        0.0       0.0   
2             0.0      0.0     0.0    0.0      0.0        0.0       0.0   
3             0.0      0.0     0.0    0.0      0.0        0.0       0.0   
5             0.0      0.0     0.0    0.0      0.0        0.0       0.0   
7             0.0      0.0     0.0    0.0      0.0        0.0       0.0   

       aesthetic  affected    afford  afraid  after  afternoon  agent  ago  \
Sr_No                                                                        
1            0.0       0.0  0.000000     0.0    0.0        0.0    0.0  0.0   
2            0.0       0.0  0.194899     0.0    0.0        0.0    0.0  0.0   
3            0.0       0.0  0.000000     0.0    0.0        0.0    0.0  0.0   
5            0.0       0.0  0.000000     0.0    0.0        0.0    0.0  0.0   
7            0.0       0.0  0.000000     0.0    0.0        0.0    0.0  0.0   

       agree  ahead  air  ajar  alive  allow  allude  alternatively  \
Sr_No                                                                 
1        0.0    0.0  0.0   0.0    0.0    0.0     0.0            0.0   
2        0.0    0.0  0.0   0.0    0.0    0.0     0.0            0.0   
3        0.0    0.0  0.0   0.0    0.0    0.0     0.0            0.0   
5        0.0    0.0  0.0   0.0    0.0    0.0     0.0            0.0   
7        0.0    0.0  0.0   0.0    0.0    0.0     0.0            0.0   

       altogether  aluminium  amazing  amazon  and  andro  android  androids  \
Sr_No                                                                          
1             0.0        0.0      0.0     0.0  0.0    0.0      0.0       0.0   
2             0.0        0.0 

In [4]:
# Dropping the un-necessary columns as well as Upvote and Downvote for our Experiment.
df.drop(['Unnamed: 0','Upvote','Downvote'],axis=1,inplace=True)

In [5]:
df.head()

1080p  12mp  1960mah  32gb  3gs  44000rs  500rs  59k  720p  7mp  8pm  \
Sr_No                                                                         
1        0.0   0.0      0.0   0.0  0.0      0.0    0.0  0.0   0.0  0.0  0.0   
2        0.0   0.0      0.0   0.0  0.0      0.0    0.0  0.0   0.0  0.0  0.0   
3        0.0   0.0      0.0   0.0  0.0      0.0    0.0  0.0   0.0  0.0  0.0   
5        0.0   0.0      0.0   0.0  0.0      0.0    0.0  0.0   0.0  0.0  0.0   
7        0.0   0.0      0.0   0.0  0.0      0.0    0.0  0.0   0.0  0.0  0.0   

         a  a10  aa10  able  absence  absolutely  accelerated  accept  \
Sr_No                                                                   
1      0.0  0.0   0.0   0.0      0.0         0.0          0.0     0.0   
2      0.0  0.0   0.0   0.0      0.0         0.0          0.0     0.0   
3      0.0  0.0   0.0   0.0      0.0         0.0          0.0     0.0   
5      0.0  0.0   0.0   0.0      0.0         0.0          0.0     0.0   
7      0.0  0.0   0.0   0.0      0.0         0.0          0.0     0.0   

       accesorie  access  accessory  accord  accumulation  achieved  \
Sr_No                                                                 
1            0.0     0.0        0.0     0.0           0.0       0.0   
2            0.0     0.0        0.0     0.0           0.0       0.0   
3            0.0     0.0        0.0     0.0           0.0       0.0   
5            0.0     0.0        0.0     0.0           0.0       0.0   
7            0.0     0.0        0.0     0.0           0.0       0.0   

       acknowledge  act  activate  activity    actual   ad  adapt  add  \
Sr_No                                                                    
1              0.0  0.0       0.0       0.0  0.000000  0.0    0.0  0.0   
2              0.0  0.0       0.0       0.0  0.194899  0.0    0.0  0.0   
3              0.0  0.0       0.0       0.0  0.000000  0.0    0.0  0.0   
5              0.0  0.0       0.0       0.0  0.000000  0.0    0.0  0.0   
7              0.0  0.0       0.0       0.0  0.000000  0.0    0.0  0.0   

       additional  address  adjust  admit  advance  advantage  advocate  \
Sr_No                                                                     
1             0.0      0.0     0.0    0.0      0.0        0.0       0.0   
2             0.0      0.0     0.0    0.0      0.0        0.0       0.0   
3             0.0      0.0     0.0    0.0      0.0        0.0       0.0   
5             0.0      0.0     0.0    0.0      0.0        0.0       0.0   
7             0.0      0.0     0.0    0.0      0.0        0.0       0.0   

       aesthetic  affected    afford  afraid  after  afternoon  agent  ago  \
Sr_No                                                                        
1            0.0       0.0  0.000000     0.0    0.0        0.0    0.0  0.0   
2            0.0       0.0  0.194899     0.0    0.0        0.0    0.0  0.0   
3            0.0       0.0  0.000000     0.0    0.0        0.0    0.0  0.0   
5            0.0       0.0  0.000000     0.0    0.0        0.0    0.0  0.0   
7            0.0       0.0  0.000000     0.0    0.0        0.0    0.0  0.0   

       agree  ahead  air  ajar  alive  allow  allude  alternatively  \
Sr_No                                                                 
1        0.0    0.0  0.0   0.0    0.0    0.0     0.0            0.0   
2        0.0    0.0  0.0   0.0    0.0    0.0     0.0            0.0   
3        0.0    0.0  0.0   0.0    0.0    0.0     0.0            0.0   
5        0.0    0.0  0.0   0.0    0.0    0.0     0.0            0.0   
7        0.0    0.0  0.0   0.0    0.0    0.0     0.0            0.0   

       altogether  aluminium  amazing  amazon  and  andro  android  androids  \
Sr_No                                                                          
1             0.0        0.0      0.0     0.0  0.0    0.0      0.0       0.0   
2             0.0        0.0      0.0     0.0  0.0    0.0      0.0       0.0   
3             0.0        0.0     

In [6]:
df.shape

(237, 1625)

### Splitting into Train and Test data

In [7]:
# All predictors in X
X=df.iloc[:,:-1]
# The target variables 'h' in y
y=df.iloc[:,-1]

In [8]:
X.head()

1080p  12mp  1960mah  32gb  3gs  44000rs  500rs  59k  720p  7mp  8pm  \
Sr_No                                                                         
1        0.0   0.0      0.0   0.0  0.0      0.0    0.0  0.0   0.0  0.0  0.0   
2        0.0   0.0      0.0   0.0  0.0      0.0    0.0  0.0   0.0  0.0  0.0   
3        0.0   0.0      0.0   0.0  0.0      0.0    0.0  0.0   0.0  0.0  0.0   
5        0.0   0.0      0.0   0.0  0.0      0.0    0.0  0.0   0.0  0.0  0.0   
7        0.0   0.0      0.0   0.0  0.0      0.0    0.0  0.0   0.0  0.0  0.0   

         a  a10  aa10  able  absence  absolutely  accelerated  accept  \
Sr_No                                                                   
1      0.0  0.0   0.0   0.0      0.0         0.0          0.0     0.0   
2      0.0  0.0   0.0   0.0      0.0         0.0          0.0     0.0   
3      0.0  0.0   0.0   0.0      0.0         0.0          0.0     0.0   
5      0.0  0.0   0.0   0.0      0.0         0.0          0.0     0.0   
7      0.0  0.0   0.0   0.0      0.0         0.0          0.0     0.0   

       accesorie  access  accessory  accord  accumulation  achieved  \
Sr_No                                                                 
1            0.0     0.0        0.0     0.0           0.0       0.0   
2            0.0     0.0        0.0     0.0           0.0       0.0   
3            0.0     0.0        0.0     0.0           0.0       0.0   
5            0.0     0.0        0.0     0.0           0.0       0.0   
7            0.0     0.0        0.0     0.0           0.0       0.0   

       acknowledge  act  activate  activity    actual   ad  adapt  add  \
Sr_No                                                                    
1              0.0  0.0       0.0       0.0  0.000000  0.0    0.0  0.0   
2              0.0  0.0       0.0       0.0  0.194899  0.0    0.0  0.0   
3              0.0  0.0       0.0       0.0  0.000000  0.0    0.0  0.0   
5              0.0  0.0       0.0       0.0  0.000000  0.0    0.0  0.0   
7              0.0  0.0       0.0       0.0  0.000000  0.0    0.0  0.0   

       additional  address  adjust  admit  advance  advantage  advocate  \
Sr_No                                                                     
1             0.0      0.0     0.0    0.0      0.0        0.0       0.0   
2             0.0      0.0     0.0    0.0      0.0        0.0       0.0   
3             0.0      0.0     0.0    0.0      0.0        0.0       0.0   
5             0.0      0.0     0.0    0.0      0.0        0.0       0.0   
7             0.0      0.0     0.0    0.0      0.0        0.0       0.0   

       aesthetic  affected    afford  afraid  after  afternoon  agent  ago  \
Sr_No                                                                        
1            0.0       0.0  0.000000     0.0    0.0        0.0    0.0  0.0   
2            0.0       0.0  0.194899     0.0    0.0        0.0    0.0  0.0   
3            0.0       0.0  0.000000     0.0    0.0        0.0    0.0  0.0   
5            0.0       0.0  0.000000     0.0    0.0        0.0    0.0  0.0   
7            0.0       0.0  0.000000     0.0    0.0        0.0    0.0  0.0   

       agree  ahead  air  ajar  alive  allow  allude  alternatively  \
Sr_No                                                                 
1        0.0    0.0  0.0   0.0    0.0    0.0     0.0            0.0   
2        0.0    0.0  0.0   0.0    0.0    0.0     0.0            0.0   
3        0.0    0.0  0.0   0.0    0.0    0.0     0.0            0.0   
5        0.0    0.0  0.0   0.0    0.0    0.0     0.0            0.0   
7        0.0    0.0  0.0   0.0    0.0    0.0     0.0            0.0   

       altogether  aluminium  amazing  amazon  and  andro  android  androids  \
Sr_No                                                                          
1             0.0        0.0      0.0     0.0  0.0    0.0      0.0       0.0   
2             0.0        0.0      0.0     0.0  0.0    0.0      0.0       0.0   
3             0.0        0.0     

In [9]:
y.head()

Sr_No
1    0.89
2    0.87
3    0.87
5    0.86
7    0.84
Name: h, dtype: float64

In [10]:
# Splitting into train and test
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=50,random_state=0)

In [11]:
X_train.shape,X_test.shape

((187, 1624), (50, 1624))

In [12]:
y_train.shape,y_test.shape

((187,), (50,))

### Random Forest Regressor: Modelling the train data

In [13]:
# Import the model
from sklearn.ensemble import RandomForestRegressor

In [14]:
# Training the Random Forest Regressor
rf = RandomForestRegressor(n_estimators = 100, random_state = 0)
rf.fit(X_train, y_train)

# Getting the feature importance
feature_importances_df = pd.DataFrame({"feature": X_train.columns, "importance": rf.feature_importances_}).sort_values("importance", ascending = False)
feature_importances_df.head(20)

feature  importance
1612      Len_before    0.202565
1611  Sum_of_Up_Down    0.119685
1619       Len_after    0.065309
1613    Num_Sentence    0.060054
1621       Perc_Verb    0.053168
1565            well    0.026585
1623        Perc_Adj    0.026213
175              buy    0.023693
1620       Perc_Noun    0.021660
1053           phone    0.018287
181           camera    0.015703
1095           price    0.012508
1622     Perc_Adverb    0.011450
561         flipkart    0.011377
1610   Review_Rating    0.011341
610              get    0.011045
1178         replace    0.009717
1487               u    0.009659
670             help    0.009378
86               ask    0.009351

### Predicing on Test Data

In [15]:
# Predicting on test data
y_pred = rf.predict(X_test)

In [16]:
y_pred

array([0.3107, 0.3857, 0.7973, 0.6486, 0.5941, 0.3479, 0.572 , 0.5168,
       0.7935, 0.3771, 0.635 , 0.6306, 0.3787, 0.6293, 0.3875, 0.4428,
       0.3812, 0.6261, 0.4822, 0.7914, 0.6604, 0.4607, 0.4455, 0.7523,
       0.4466, 0.5037, 0.3366, 0.7874, 0.5549, 0.567 , 0.5782, 0.6423,
       0.5536, 0.3453, 0.3452, 0.4187, 0.277 , 0.3762, 0.7312, 0.6094,
       0.6685, 0.622 , 0.4459, 0.6693, 0.5096, 0.7611, 0.7405, 0.3754,
       0.5895, 0.4176])

In [17]:
# Creating a dataframe to compare predicted and test values
pd.DataFrame({"y_pred": y_pred, "y_test": y_test})

y_pred  y_test
Sr_No                
354    0.3107    0.45
304    0.3857    0.57
144    0.7973    0.65
150    0.6486    0.52
193    0.5941    0.47
1085   0.3479    0.31
189    0.5720    0.54
360    0.5168    0.36
57     0.7935    0.92
1046   0.3771    0.36
185    0.6350    0.73
1032   0.6306    0.40
1086   0.3787    0.29
139    0.6293    0.55
1259   0.3875    0.10
305    0.4428    0.58
379    0.3812    0.43
146    0.6261    0.63
366    0.4822    0.42
13     0.7914    0.87
72     0.6604    0.64
124    0.4607    0.68
120    0.4455    0.74
8      0.7523    0.84
29     0.4466    0.84
49     0.5037    0.79
229    0.3366    0.60
19     0.7874    0.82
147    0.5549    0.53
220    0.5670    0.43
1075   0.5782    0.25
155    0.6423    0.49
1081   0.5536    0.33
364    0.3453    0.40
56     0.3452    0.92
302    0.4187    0.50
346    0.2770    0.50
1079   0.3762    0.38
97     0.7312    0.71
131    0.6094    0.65
94     0.6685    0.71
25     0.6220    0.76
301    0.4459    0.45
22     0.6693    0.81
83     0.5096    0.68
12     0.7611    0.84
298    0.7405    0.52
351    0.3754    0.40
228    0.5895    0.60
380    0.4176    0.45

### Calculating Mean Absolute Percentage Error (MAPE)

In [18]:
# Mean Absolute Percentage Error (MAPE)
mape = np.mean(np.abs((y_test - y_pred) / y_test)) * 100

In [19]:
mape

28.894314086100593

### XGBoost Regressor: Modelling the train data

In [21]:
from xgboost import XGBRegressor

In [22]:
# Training the Random Forest Regressor
xgb = XGBRegressor(n_estimators = 100, random_state = 0)
xgb.fit(X_train, y_train)

XGBRegressor(base_score=0.5, booster=None, colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
             importance_type='gain', interaction_constraints=None,
             learning_rate=0.300000012, max_delta_step=0, max_depth=6,
             min_child_weight=1, missing=nan, monotone_constraints=None,
             n_estimators=100, n_jobs=0, num_parallel_tree=1,
             objective='reg:squarederror', random_state=0, reg_alpha=0,
             reg_lambda=1, scale_pos_weight=1, subsample=1, tree_method=None,
             validate_parameters=False, verbosity=None)

### Predicting on Test Data

In [23]:
# Predicting on test data
y_pred = xgb.predict(X_test)

In [24]:
y_pred

array([0.292058  , 0.3437006 , 0.803192  , 0.6926938 , 0.6649079 ,
       0.41608042, 0.5249463 , 0.5553055 , 0.7665918 , 0.40633464,
       0.5729065 , 0.6731767 , 0.39362898, 0.622809  , 0.23668745,
       0.4711731 , 0.27968174, 0.56039345, 0.44363052, 0.8034617 ,
       0.6321041 , 0.31824607, 0.41848937, 0.7810065 , 0.504666  ,
       0.43484944, 0.37464938, 0.7477822 , 0.46644872, 0.5513485 ,
       0.5198137 , 0.5350493 , 0.633767  , 0.37416667, 0.34359312,
       0.42001548, 0.21934605, 0.34956646, 0.7407421 , 0.51067203,
       0.607087  , 0.5825378 , 0.53317636, 0.71797025, 0.43868402,
       0.7703686 , 0.6629163 , 0.34871525, 0.6585605 , 0.31032324],
      dtype=float32)

In [25]:
# Creating a dataframe to compare predicted and test values
pd.DataFrame({"y_pred": y_pred, "y_test": y_test})

y_pred  y_test
Sr_No                  
354    0.292058    0.45
304    0.343701    0.57
144    0.803192    0.65
150    0.692694    0.52
193    0.664908    0.47
1085   0.416080    0.31
189    0.524946    0.54
360    0.555305    0.36
57     0.766592    0.92
1046   0.406335    0.36
185    0.572906    0.73
1032   0.673177    0.40
1086   0.393629    0.29
139    0.622809    0.55
1259   0.236687    0.10
305    0.471173    0.58
379    0.279682    0.43
146    0.560393    0.63
366    0.443631    0.42
13     0.803462    0.87
72     0.632104    0.64
124    0.318246    0.68
120    0.418489    0.74
8      0.781007    0.84
29     0.504666    0.84
49     0.434849    0.79
229    0.374649    0.60
19     0.747782    0.82
147    0.466449    0.53
220    0.551349    0.43
1075   0.519814    0.25
155    0.535049    0.49
1081   0.633767    0.33
364    0.374167    0.40
56     0.343593    0.92
302    0.420015    0.50
346    0.219346    0.50
1079   0.349566    0.38
97     0.740742    0.71
131    0.510672    0.65
94     0.607087    0.71
25     0.582538    0.76
301    0.533176    0.45
22     0.717970    0.81
83     0.438684    0.68
12     0.770369    0.84
298    0.662916    0.52
351    0.348715    0.40
228    0.658561    0.60
380    0.310323    0.45

### Calculating Mean Absolute Percentage Error (MAPE)

In [26]:
# Mean Absolute Percentage Error (MAPE)
mape = np.mean(np.abs((y_test - y_pred) / y_test)) * 100

In [27]:
mape

30.009617615495763

#### When we drop Upvote and Downvote columns, fit our model on train data and try to test on test data, we see that MAPE increases considerably as compared when these two columns were included.
#### We got MAPE of 28.89 and 30 using Random Forest and XGBoost respectively excluding Upvote and Downvote. When Upvote and Downvote was included we got MAPE of 17.26 and 16.47 using Random Forest and XGBoost respectively.